pkg/retriever/bitswapretriever.go

```diff
@@ -166,15 +181,50 @@ func (br *bitswapRetrieval) RetrieveFromAsyncCandidates(ayncCandidates types.Inb
 			br.routing.AddProviders(br.request.RetrievalID, nextCandidates)
 		}
 	}()
+
+	// loader that can work with bitswap just for this session
+	loader := loaderForSession(br.request.RetrievalID, br.inProgressCids, br.bsGetter)
+
+	// set up the storage system, including the preloader
+	concurrency := br.BitswapRetriever.cfg.Concurrency
+	if concurrency <= 0 {
+		concurrency = DefaultConcurrency
+	}
+	cacheLinkSys := br.request.PreloadLinkSystem
+	if cacheLinkSys.StorageReadOpener == nil || cacheLinkSys.StorageWriteOpener == nil {
+		// an uninitialised LinkSystem that we can't actually use, so we'll create a
+		// new one
+		tmpDir := br.BitswapRetriever.cfg.TempDir
+		if tmpDir == "" {
+			tmpDir = os.TempDir()
+		}
+		cacheStore := storage.NewDeferredCarStorage(tmpDir)
```

Comment:
could the preload cache be fully in memory? presumably we don't need _that_ much of a preload per active retrieval and by keeping those working sets in memory we save additional disk latencies

Reply:
Yes .. so this started off as in-memory (it's based on @hannahhoward's fully in-memory buffering preload store in the go-ipld-prime PR) but the concern is that there are real cases where a preloader will fill up memory, especially in the Saturn case where you're doing many parallel fetches. e.g. plain unixfs shard with 256 links and you only care about 1 initially, so you go down the next, preloader works on the other 255, the next jump has 256 more, etc. onward down until you come back up and start chewing away at the next branch; that could compound quickly with many parallel fetches.

Future work for this general preloading approach I imagined would have both an in-memory and a filesystem backed storage, where your in-memory objects can overflow into the filesystem if you end up holding too much. The vast majority of cases will not need more than a couple of blocks from the preloader so this would speed things up significantly.

But there's another issue that gets in the way - the traverser needs access to blocks its already passed for those DAG cases where you re-encounter blocks via different paths. They're less common with UnixFS but can happen, and we're also not limited to UnixFS here either. SO that means we need to hang onto the full suite of blocks for the life of a retrieval anyway. Using the approach I've gone with here we just put them all into the same basket on disk, using a CAR as a temporary store, not caring about the ordering, so the blocks are available to the preloader or the loader and in the case of a re-load via the loader. But it's also true that this could be done in memory up to some limit! So I'd really like to work on a storage system that starts in memory but overflows to disk, which we can then use for all the cases we need it in Lassie - graphsync, bitswap, preloader. The user only ever gets a nicely ordered CARv1 and we get to mess around with different ways of juggling blocks internally.

Reply:
For the selector we're using here do we actually need block data on re-encounter, or would it be sufficient to keep just the list of encountered CIDs so that we can properly answer 'Has' to remember which ones have already been interacted with?

Reply:
For the ones we're using I think we probably could .. we do allow for arbitrary selectors in the retrieval `Request` that can be passed in when using Lassie as a library, and this is yet again one of those cases where it'd be nice to be able to differentiate between selectors that care about revisits and those that don't.

For the main cases we actually care about (these simple path-based selectors), we could probably get away with just a `SkipMe{}` whenever the LinkSystem is asked for a block that it's already encountered. In which case, just maintaining a list of CIDs would work. @hannahhoward is there anything in the graphsync traverser that would complicate this kind of approach (if we were hypothetically able to restrict the types of selectors we're willing to work with)?

I think maybe not in scope for this batch of work, but we could explore options for the "have simple selector, won't store anything but a list of CIDs we've already spat out" case later.